crates/turborepo-lib/src/run/summary/mod.rs

```diff
@@ -0,0 +1,154 @@
+mod execution;
+mod global_hash;
+mod scm;
+mod spaces;
+mod task;
+
+use global_hash::GlobalHashSummary;
+use serde::{Deserialize, Serialize};
+use svix_ksuid::Ksuid;
+use thiserror::Error;
+use turbopath::{
+    AbsoluteSystemPath, AbsoluteSystemPathBuf, AnchoredSystemPath, AnchoredSystemPathBuf,
+};
+use turborepo_api_client::APIClient;
+use turborepo_ci::Vendor;
+use turborepo_env::EnvironmentVariableMap;
+
+use crate::{
+    cli::EnvMode,
+    opts::RunOpts,
+    run::summary::{execution::ExecutionSummary, scm::SCMState, task::TaskSummary},
+};
+
+#[derive(Debug, Error)]
+enum Error {}
+
+// NOTE: When changing this, please ensure that the server side is updated to
+// handle the new version on vercel.com this is required to ensure safe handling
+// of env vars (unknown run summary versions will be ignored on the server)
+const RUN_SUMMARY_SCHEMA_VERSION: &str = "1";
+
+enum RunType {
+    Real,
+    DryText,
+    DryJson,
+}
+
+fn get_user(env_vars: &EnvironmentVariableMap) -> Option<String> {
+    if turborepo_ci::is_ci() {
+        return Vendor::get_info()
+            .and_then(|vendor| vendor.username_env_var)
+            .and_then(|username_env_var| env_vars.get(username_env_var).cloned());
+    }
+
+    None
+}
+
+// Wrapper around the serializable RunSummary, with some extra information
+// about the Run and references to other things that we need.
+struct Meta<'a> {
+    run_summary: RunSummary<'a>,
+    repo_root: &'a AbsoluteSystemPath,
+    single_package: bool,
+    should_save: bool,
+    run_type: RunType,
+    synthesized_command: String,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct RunSummary<'a> {
+    id: Ksuid,
+    version: String,
+    turbo_version: String,
+    monorepo: bool,
+    global_hash_summary: GlobalHashSummary,
+    packages: Vec<String>,
+    env_mode: EnvMode,
+    framework_inference: bool,
+    execution_summary: Option<ExecutionSummary<'a>>,
+    tasks: Vec<TaskSummary>,
```

Comment:
I think this is the only area that we need to make a decision on. In Go we [essentially create](https://github.com/vercel/turbo/blob/main/cli/internal/run/real_run.go#L209) an `Arc<Mutex<Vec<TaskSummary>>>` in the task graph visitor and append to it after each task finishes. In the visitor we also ["close" the task](https://github.com/vercel/turbo/blob/main/cli/internal/run/real_run.go#L257) which just sends it to spaces if that feature is enabled. I think it might make sense to go with a `mpsc` pattern and have the consumer update the state and also send off the info to spaces in the background.